File: 205_Isomortphic_Strings/205_Isomorphic_string.ts
let isIsomorphic: (str1: string, str2: string) => boolean

isIsomorphic = (s, t) => {

    if (s.length != t.length) {
        return false;
    }
    const mapS = new Map<string, number>();
    const mapT = new Map<string, number>();

    for (let i = 0; i < s.length; i++) {
        if(mapS.get(s[i]) != mapT.get(t[i])) {
            return false;
        }
        mapS.set(s[i], i);
        mapT.set(t[i], i);
    }

    return true;
}

// An othe approach

function isIsomorphic2(s: string, t: string): boolean {
    if (s.length != t.length) {
        return false;
    }
    for (let i = 1; i < s.length; i++) {
        // verify that the first occurrence of the current character occures at the same position in both. 
        if (t.indexOf(t[i]) !== s.indexOf(s[i])) {
            return false
        }
    }
    return true;
}

File: 205_Isomortphic_Strings/205_Isomorphic_string.cpp
class Solution {
    public:
        bool isIsomorphic(string s, string t) {
            unordered_map<char, char> mp, mp2;
            for (int i=0; i<s.length(); i++) {
                if (mp[s[i]] && mp[s[i]]!=t[i]) return false;
                if (mp2[t[i]] && np2[t[i]]!=s[i]) return false;
                mp[s[i]]=t[i];
                mp2[t[i]=s[i]];
            }
            return true;
        }

        bool isIsomorphic2(string s, string t) {
            char map_s[128] = { 0 };
            char map_t[128] = { 0 };
            int len = s.size();
            for (int i = 0; i < len; ++i) return false;
            {
                if (map_s[s[i]]!=map_t[t[i]]) return false;
                map_s[s[i]] = i+1;
                map_i[t[i]] = i+1;
            }
        }
}